{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import os, sys, glob, re, math, pickle\n",
    "import phate, scprep, magic, meld\n",
    "import graphtools as gt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import time,random,datetime\n",
    "import networkx as nx\n",
    "import scvelo as scv\n",
    "from sklearn import metrics\n",
    "from sklearn import model_selection\n",
    "from scipy import sparse\n",
    "from scipy.stats import mannwhitneyu, tiecorrect, rankdata\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import scanpy as sc\n",
    "from sklearn.dummy import DummyClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.cluster import SpectralClustering, OPTICS, cluster_optics_dbscan, AgglomerativeClustering\n",
    "from bbknn import bbknn\n",
    "import warnings\n",
    "%matplotlib inline\n",
    "%load_ext memory_profiler\n",
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "import anndata2ri\n",
    "\n",
    "# Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "# settings\n",
    "plt.rc('font', size = 8)\n",
    "plt.rc('font', family='sans serif')\n",
    "plt.rcParams['pdf.fonttype']=42\n",
    "plt.rcParams['ps.fonttype']=42\n",
    "plt.rcParams['text.usetex']=False\n",
    "plt.rcParams['legend.frameon']=False\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['legend.markerscale']=0.5\n",
    "sc.set_figure_params(dpi=300,dpi_save=600,\n",
    "                     frameon=False,\n",
    "                     fontsize=8)\n",
    "plt.rcParams['savefig.dpi']=600\n",
    "sc.settings.verbosity=2\n",
    "sc._settings.ScanpyConfig.n_jobs=-1\n",
    "\n",
    "# reproducibility\n",
    "rs = np.random.seed(42)\n",
    "\n",
    "# utils\n",
    "def mwu(X,Y,gene_names,correction=None,debug=False) :\n",
    "    '''\n",
    "    Benjamini-Hochberg correction implemented. Can change to Bonferonni\n",
    "    gene_names (list)\n",
    "    if X,Y single gene expression array, input x.reshape(-1,1), y.reshape(-1,1)\n",
    "    NOTE: get zeros sometimes because difference (p-value is so small)\n",
    "    '''\n",
    "    p=pd.DataFrame()\n",
    "    print('Mann-Whitney U w/Benjamini/Hochberg correction\\n')\n",
    "    start = time.time()\n",
    "    for i,g in enumerate(gene_names) :\n",
    "        if i==np.round(np.quantile(np.arange(len(gene_names)),0.25)) :\n",
    "            print('... 25% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.5)) :\n",
    "            print('... 50% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        elif i==np.round(np.quantile(np.arange(len(gene_names)),0.75)) :\n",
    "            print('... 75% completed in {:.2f}-s'.format(time.time()-start))\n",
    "        p.loc[i,'Gene']=g\n",
    "        if (tiecorrect(rankdata(np.concatenate((np.asarray(X[:,i]),np.asarray(Y[:,i])))))==0) :\n",
    "            if debug :\n",
    "                print('P-value not calculable for {}'.format(g))\n",
    "            p.loc[i,'pval']=np.nan\n",
    "        else :\n",
    "            _,p.loc[i,'pval']=mannwhitneyu(X[:,i],Y[:,i]) # continuity correction is True\n",
    "    print('... mwu computed in {:.2f}-s\\n'.format(time.time() - start))\n",
    "    # ignore NaNs, since can't do a comparison on these (change numbers for correction)\n",
    "    p_corrected = p.loc[p['pval'].notna(),:]\n",
    "    new_pvals = multipletests(p_corrected['pval'],method='fdr_bh')\n",
    "    p_corrected['pval_corrected'] = new_pvals[1]\n",
    "    return p_corrected\n",
    "\n",
    "def log2aveFC(X,Y,gene_names,AnnData=None) :\n",
    "    '''not sensitivity to directionality due to subtraction\n",
    "    X and Y full arrays, subsetting performed here\n",
    "    `gene_names` (list): reduced list of genes to calc\n",
    "    `adata` (sc.AnnData): to calculate reduced list. NOTE: assumes X,Y drawn from adata.var_names\n",
    "    '''\n",
    "    if not AnnData is None :\n",
    "        g_idx = [i for i,g in enumerate(AnnData.var_names) if g in gene_names]\n",
    "        fc=pd.DataFrame({'Gene':AnnData.var_names[g_idx],\n",
    "                         'log2FC':np.log2(X[:,g_idx].mean(axis=0)) - np.log2(Y[:,g_idx].mean(axis=0))}) # returns NaN if negative value \n",
    "    else :\n",
    "        fc=pd.DataFrame({'Gene':gene_names,\n",
    "                         'log2FC':np.log2(X.mean(axis=0)) - np.log2(Y.mean(axis=0))})\n",
    "    return fc\n",
    "\n",
    "\n",
    "# fps\n",
    "dfp = '/home/cl2292/project/SBMA/26wk/data/'\n",
    "pfp = '/home/cl2292/project/SBMA/26wk/results/'\n",
    "pdfp = '/home/cl2292/project/SBMA/26wk/data/processed/'\n",
    "sc.settings.figdir = pfp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak memory: 16016.39 MiB, increment: 15716.36 MiB\n",
      "loaded @220224.13:01:40\n",
      "took 36.13-s to load data\n"
     ]
    }
   ],
   "source": [
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='220223_subcluster.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit adata = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WT 26wk\n",
      "\n",
      " OL     3050\n",
      "NU     2512\n",
      "AS2     927\n",
      "MG      305\n",
      "AS1     251\n",
      "OPC     227\n",
      "END     161\n",
      "EP       48\n",
      "MN       38\n",
      "CN       20\n",
      "PER      16\n",
      "Name: sub3, dtype: int64 \n",
      "\n",
      "\n",
      "WT 52wk\n",
      "\n",
      " OL     8140\n",
      "NU     7778\n",
      "AS2    3325\n",
      "AS1     920\n",
      "OPC     748\n",
      "MG      738\n",
      "END     358\n",
      "EP       94\n",
      "CN       75\n",
      "PER      23\n",
      "MN       18\n",
      "Name: sub3, dtype: int64 \n",
      "\n",
      "\n",
      "AR 26wk\n",
      "\n",
      " OL     2942\n",
      "NU     2391\n",
      "AS2     980\n",
      "MG      251\n",
      "OPC     224\n",
      "AS1     217\n",
      "END     118\n",
      "EP       42\n",
      "MN       24\n",
      "CN       24\n",
      "PER      17\n",
      "Name: sub3, dtype: int64 \n",
      "\n",
      "\n",
      "AR 52wk\n",
      "\n",
      " OL     9621\n",
      "NU     8038\n",
      "AS2    3607\n",
      "AS1    1156\n",
      "MG      968\n",
      "OPC     777\n",
      "END     377\n",
      "EP      187\n",
      "CN       80\n",
      "MN       50\n",
      "PER      21\n",
      "Name: sub3, dtype: int64 \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## # of cell types in each timepoint&genotype\n",
    "print('WT 26wk\\n\\n',adata[(adata.obs['genotype']=='WT')&(adata.obs['timepoint']=='26wk'),:].obs['sub3'].value_counts(), '\\n\\n')\n",
    "print('WT 52wk\\n\\n',adata[(adata.obs['genotype']=='WT')&(adata.obs['timepoint']=='52wk'),:].obs['sub3'].value_counts(), '\\n\\n')\n",
    "print('AR 26wk\\n\\n',adata[(adata.obs['genotype']=='AR')&(adata.obs['timepoint']=='26wk'),:].obs['sub3'].value_counts(), '\\n\\n')\n",
    "print('AR 52wk\\n\\n',adata[(adata.obs['genotype']=='AR')&(adata.obs['timepoint']=='52wk'),:].obs['sub3'].value_counts(), '\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52wk    22217\n",
       "26wk     7555\n",
       "Name: timepoint, dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[(adata.obs['genotype']=='WT'),:].obs['timepoint'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52wk    24882\n",
       "26wk     7230\n",
       "Name: timepoint, dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[(adata.obs['genotype']=='AR'),:].obs['timepoint'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak memory: 29748.84 MiB, increment: 13742.96 MiB\n",
      "loaded @220224.13:02:15\n",
      "took 34.65-s to load data\n",
      "peak memory: 44572.57 MiB, increment: 14823.74 MiB\n",
      "loaded @220224.13:02:51\n",
      "took 35.65-s to load data\n"
     ]
    }
   ],
   "source": [
    "# Load Data\n",
    "\n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='220224_WT_imp.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit wt = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))\n",
    "    \n",
    "if True :\n",
    "    start = time.time()\n",
    "    backed=None # None if not\n",
    "    fname='220224_SBMA_imp.h5ad' # for full, can maybe get away with ~300G\n",
    "    %memit mut = sc.read_h5ad(os.path.join(pdfp,fname),backed=backed)\n",
    "    print('loaded @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))\n",
    "    print('took {:.2f}-s to load data'.format(time.time()-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating 26wk\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "OL, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:3050\n",
      "    Ncells in Y:2942\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 28.81-s\n",
      "... 50% completed in 60.39-s\n",
      "... 75% completed in 95.94-s\n",
      "... mwu computed in 135.56-s\n",
      "\n",
      "... computed in 156.29-s\n",
      "\n",
      "AS2, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:927\n",
      "    Ncells in Y:980\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 20.04-s\n",
      "... 50% completed in 43.16-s\n",
      "... 75% completed in 70.35-s\n",
      "... mwu computed in 101.53-s\n",
      "\n",
      "... computed in 108.35-s\n",
      "\n",
      "NU, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:2512\n",
      "    Ncells in Y:2391\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 27.74-s\n",
      "... 50% completed in 58.60-s\n",
      "... 75% completed in 93.40-s\n",
      "... mwu computed in 132.43-s\n",
      "\n",
      "... computed in 145.71-s\n",
      "\n",
      "OPC, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:227\n",
      "    Ncells in Y:224\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.80-s\n",
      "... 50% completed in 36.84-s\n",
      "... 75% completed in 61.19-s\n",
      "... mwu computed in 89.82-s\n",
      "\n",
      "... computed in 93.69-s\n",
      "\n",
      "MG, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:305\n",
      "    Ncells in Y:251\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 17.07-s\n",
      "... 50% completed in 37.30-s\n",
      "... 75% completed in 61.57-s\n",
      "... mwu computed in 89.76-s\n",
      "\n",
      "... computed in 93.98-s\n",
      "\n",
      "MN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:38\n",
      "    Ncells in Y:24\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.80-s\n",
      "... 50% completed in 34.78-s\n",
      "... 75% completed in 57.70-s\n",
      "... mwu computed in 84.58-s\n",
      "\n",
      "... computed in 87.84-s\n",
      "\n",
      "AS1, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:251\n",
      "    Ncells in Y:217\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.99-s\n",
      "... 50% completed in 37.19-s\n",
      "... 75% completed in 61.40-s\n",
      "... mwu computed in 89.64-s\n",
      "\n",
      "... computed in 93.70-s\n",
      "\n",
      "END, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:161\n",
      "    Ncells in Y:118\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.25-s\n",
      "... 50% completed in 35.74-s\n",
      "... 75% completed in 59.26-s\n",
      "... mwu computed in 86.73-s\n",
      "\n",
      "... computed in 90.32-s\n",
      "\n",
      "CN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:20\n",
      "    Ncells in Y:24\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.17-s\n",
      "... 50% completed in 35.52-s\n",
      "... 75% completed in 58.86-s\n",
      "... mwu computed in 86.14-s\n",
      "\n",
      "... computed in 89.47-s\n",
      "\n",
      "PER, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:16\n",
      "    Ncells in Y:17\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.55-s\n",
      "... 50% completed in 34.40-s\n",
      "... 75% completed in 57.19-s\n",
      "... mwu computed in 83.97-s\n",
      "\n",
      "... computed in 87.16-s\n",
      "\n",
      "EP, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:48\n",
      "    Ncells in Y:42\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.94-s\n",
      "... 50% completed in 35.00-s\n",
      "... 75% completed in 58.07-s\n",
      "... mwu computed in 85.15-s\n",
      "\n",
      "... computed in 88.55-s\n",
      "\n",
      "Finished timepoint 26wk in 18.92-min\n",
      "Evaluating 52wk\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "OL, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:8140\n",
      "    Ncells in Y:9621\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 58.49-s\n",
      "... 50% completed in 119.41-s\n",
      "... 75% completed in 184.34-s\n",
      "... mwu computed in 253.78-s\n",
      "\n",
      "... computed in 298.39-s\n",
      "\n",
      "AS2, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:3325\n",
      "    Ncells in Y:3607\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 32.12-s\n",
      "... 50% completed in 67.15-s\n",
      "... 75% completed in 105.98-s\n",
      "... mwu computed in 149.20-s\n",
      "\n",
      "... computed in 167.07-s\n",
      "\n",
      "NU, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:7778\n",
      "    Ncells in Y:8038\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 56.57-s\n",
      "... 50% completed in 115.74-s\n",
      "... 75% completed in 179.62-s\n",
      "... mwu computed in 247.87-s\n",
      "\n",
      "... computed in 289.23-s\n",
      "\n",
      "OPC, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:748\n",
      "    Ncells in Y:777\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.29-s\n",
      "... 50% completed in 41.83-s\n",
      "... 75% completed in 68.20-s\n",
      "... mwu computed in 98.71-s\n",
      "\n",
      "... computed in 105.35-s\n",
      "\n",
      "MG, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:738\n",
      "    Ncells in Y:968\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 19.37-s\n",
      "... 50% completed in 41.95-s\n",
      "... 75% completed in 68.72-s\n",
      "... mwu computed in 99.50-s\n",
      "\n",
      "... computed in 106.18-s\n",
      "\n",
      "MN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:18\n",
      "    Ncells in Y:50\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.78-s\n",
      "... 50% completed in 34.68-s\n",
      "... 75% completed in 57.69-s\n",
      "... mwu computed in 84.66-s\n",
      "\n",
      "... computed in 87.95-s\n",
      "\n",
      "AS1, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:920\n",
      "    Ncells in Y:1156\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 20.53-s\n",
      "... 50% completed in 44.23-s\n",
      "... 75% completed in 71.84-s\n",
      "... mwu computed in 103.48-s\n",
      "\n",
      "... computed in 110.74-s\n",
      "\n",
      "END, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:358\n",
      "    Ncells in Y:377\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 17.31-s\n",
      "... 50% completed in 37.81-s\n",
      "... 75% completed in 62.44-s\n",
      "... mwu computed in 91.18-s\n",
      "\n",
      "... computed in 95.89-s\n",
      "\n",
      "CN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:75\n",
      "    Ncells in Y:80\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.36-s\n",
      "... 50% completed in 35.98-s\n",
      "... 75% completed in 59.50-s\n",
      "... mwu computed in 87.13-s\n",
      "\n",
      "... computed in 90.59-s\n",
      "\n",
      "PER, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:23\n",
      "    Ncells in Y:21\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.43-s\n",
      "... 50% completed in 34.04-s\n",
      "... 75% completed in 56.67-s\n",
      "... mwu computed in 83.38-s\n",
      "\n",
      "... computed in 86.78-s\n",
      "\n",
      "EP, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:94\n",
      "    Ncells in Y:187\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.23-s\n",
      "... 50% completed in 35.59-s\n",
      "... 75% completed in 59.11-s\n",
      "... mwu computed in 86.71-s\n",
      "\n",
      "... computed in 90.38-s\n",
      "\n",
      "Finished timepoint 52wk in 25.48-min\n",
      "DGE finished in 44.48-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB3, IMP\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub3'\n",
    "    fname = 'imp' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['26wk','52wk'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in wt.obs[group].unique() :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SBMA'.format(i))\n",
    "            print('----')\n",
    "            X = wt[((wt.obs[group]==i) & (wt.obs['timepoint']==t)), :].layers['imputed']\n",
    "            Y = mut[((mut.obs[group]==i) & (mut.obs['timepoint']==t)), :].layers['imputed']\n",
    "            \n",
    "\n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=adata.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,wt.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(X,Y,adata.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_down.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_up.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating 26wk\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "OL, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:3050\n",
      "    Ncells in Y:2942\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 21.38-s\n",
      "... 50% completed in 45.54-s\n",
      "... 75% completed in 73.86-s\n",
      "... mwu computed in 106.32-s\n",
      "\n",
      "... computed in 121.73-s\n",
      "\n",
      "AS2, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:927\n",
      "    Ncells in Y:980\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 17.72-s\n",
      "... 50% completed in 38.53-s\n",
      "... 75% completed in 63.42-s\n",
      "... mwu computed in 92.36-s\n",
      "\n",
      "... computed in 97.92-s\n",
      "\n",
      "NU, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:2512\n",
      "    Ncells in Y:2391\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 21.89-s\n",
      "... 50% completed in 46.76-s\n",
      "... 75% completed in 75.88-s\n",
      "... mwu computed in 108.92-s\n",
      "\n",
      "... computed in 118.20-s\n",
      "\n",
      "OPC, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:227\n",
      "    Ncells in Y:224\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.86-s\n",
      "... 50% completed in 34.81-s\n",
      "... 75% completed in 58.02-s\n",
      "... mwu computed in 85.41-s\n",
      "\n",
      "... computed in 89.02-s\n",
      "\n",
      "MG, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:305\n",
      "    Ncells in Y:251\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.06-s\n",
      "... 50% completed in 35.28-s\n",
      "... 75% completed in 58.53-s\n",
      "... mwu computed in 85.88-s\n",
      "\n",
      "... computed in 89.73-s\n",
      "\n",
      "MN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:38\n",
      "    Ncells in Y:24\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.23-s\n",
      "... 50% completed in 33.64-s\n",
      "... 75% completed in 56.06-s\n",
      "... mwu computed in 82.67-s\n",
      "\n",
      "... computed in 85.90-s\n",
      "\n",
      "AS1, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:251\n",
      "    Ncells in Y:217\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.17-s\n",
      "... 50% completed in 35.45-s\n",
      "... 75% completed in 58.83-s\n",
      "... mwu computed in 86.27-s\n",
      "\n",
      "... computed in 90.08-s\n",
      "\n",
      "END, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:161\n",
      "    Ncells in Y:118\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.79-s\n",
      "... 50% completed in 34.79-s\n",
      "... 75% completed in 57.78-s\n",
      "... mwu computed in 84.81-s\n",
      "\n",
      "... computed in 88.22-s\n",
      "\n",
      "CN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:20\n",
      "    Ncells in Y:24\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.71-s\n",
      "... 50% completed in 34.62-s\n",
      "... 75% completed in 57.61-s\n",
      "... mwu computed in 84.67-s\n",
      "\n",
      "... computed in 88.04-s\n",
      "\n",
      "PER, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:16\n",
      "    Ncells in Y:17\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 14.92-s\n",
      "... 50% completed in 33.00-s\n",
      "... 75% completed in 55.15-s\n",
      "... mwu computed in 81.43-s\n",
      "\n",
      "... computed in 84.64-s\n",
      "\n",
      "EP, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:48\n",
      "    Ncells in Y:42\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.40-s\n",
      "... 50% completed in 33.93-s\n",
      "... 75% completed in 56.48-s\n",
      "... mwu computed in 83.10-s\n",
      "\n",
      "... computed in 86.41-s\n",
      "\n",
      "Finished timepoint 26wk in 17.33-min\n",
      "Evaluating 52wk\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "OL, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:8140\n",
      "    Ncells in Y:9621\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 34.54-s\n",
      "... 50% completed in 71.54-s\n",
      "... 75% completed in 113.03-s\n",
      "... mwu computed in 158.94-s\n",
      "\n",
      "... computed in 184.29-s\n",
      "\n",
      "AS2, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:3325\n",
      "    Ncells in Y:3607\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 23.44-s\n",
      "... 50% completed in 49.87-s\n",
      "... 75% completed in 80.46-s\n",
      "... mwu computed in 115.17-s\n",
      "\n",
      "... computed in 127.16-s\n",
      "\n",
      "NU, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:7778\n",
      "    Ncells in Y:8038\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 36.54-s\n",
      "... 50% completed in 75.64-s\n",
      "... 75% completed in 119.18-s\n",
      "... mwu computed in 166.88-s\n",
      "\n",
      "... computed in 191.88-s\n",
      "\n",
      "OPC, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:748\n",
      "    Ncells in Y:777\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 17.44-s\n",
      "... 50% completed in 38.01-s\n",
      "... 75% completed in 62.63-s\n",
      "... mwu computed in 91.47-s\n",
      "\n",
      "... computed in 96.80-s\n",
      "\n",
      "MG, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:738\n",
      "    Ncells in Y:968\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 17.47-s\n",
      "... 50% completed in 38.05-s\n",
      "... 75% completed in 62.90-s\n",
      "... mwu computed in 91.80-s\n",
      "\n",
      "... computed in 97.33-s\n",
      "\n",
      "MN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:18\n",
      "    Ncells in Y:50\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.15-s\n",
      "... 50% completed in 33.30-s\n",
      "... 75% completed in 55.63-s\n",
      "... mwu computed in 82.10-s\n",
      "\n",
      "... computed in 85.36-s\n",
      "\n",
      "AS1, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:920\n",
      "    Ncells in Y:1156\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 18.31-s\n",
      "... 50% completed in 39.65-s\n",
      "... 75% completed in 65.19-s\n",
      "... mwu computed in 94.82-s\n",
      "\n",
      "... computed in 100.80-s\n",
      "\n",
      "END, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:358\n",
      "    Ncells in Y:377\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.78-s\n",
      "... 50% completed in 36.63-s\n",
      "... 75% completed in 60.65-s\n",
      "... mwu computed in 88.77-s\n",
      "\n",
      "... computed in 93.02-s\n",
      "\n",
      "CN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:75\n",
      "    Ncells in Y:80\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.11-s\n",
      "... 50% completed in 35.38-s\n",
      "... 75% completed in 58.69-s\n",
      "... mwu computed in 86.06-s\n",
      "\n",
      "... computed in 89.44-s\n",
      "\n",
      "PER, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:23\n",
      "    Ncells in Y:21\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.03-s\n",
      "... 50% completed in 33.24-s\n",
      "... 75% completed in 55.62-s\n",
      "... mwu computed in 82.18-s\n",
      "\n",
      "... computed in 85.56-s\n",
      "\n",
      "EP, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:94\n",
      "    Ncells in Y:187\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.74-s\n",
      "... 50% completed in 34.63-s\n",
      "... 75% completed in 57.48-s\n",
      "... mwu computed in 84.47-s\n",
      "\n",
      "... computed in 87.90-s\n",
      "\n",
      "Finished timepoint 52wk in 20.66-min\n",
      "DGE finished in 38.05-min\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB3, No IMP,\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub3'\n",
    "    fname = 'no_imp' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['26wk','52wk'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in wt.obs[group].unique() :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SBMA'.format(i))\n",
    "            print('----')\n",
    "            X = wt.X[((wt.obs[group]==i) & (wt.obs['timepoint']==t)), :]\n",
    "            Y = mut.X[((mut.obs[group]==i) & (mut.obs['timepoint']==t)), :]\n",
    "            \n",
    "            if sparse.issparse(X):\n",
    "                X = X.todense()\n",
    "            if sparse.issparse(Y):\n",
    "                Y = Y.todense()\n",
    "                \n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=adata.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,adata.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(X,Y,adata.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_down.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_up.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved @220224.13:45:19\n"
     ]
    }
   ],
   "source": [
    "## Sampling\n",
    "\n",
    "wt26 = wt.obs.loc[wt.obs['timepoint']=='26wk', :].sample(n=7230, replace=False).index.to_list()\n",
    "wt52 = wt.obs.loc[wt.obs['timepoint']=='52wk', :].sample(n=7230, replace=False).index.to_list()\n",
    "sample_wt = wt[(wt.obs.index.isin(wt26)|wt.obs.index.isin(wt52)), :]\n",
    "\n",
    "mut26 = mut.obs.loc[mut.obs['timepoint']=='26wk', :].sample(n=7230, replace=False).index.to_list()\n",
    "mut52 = mut.obs.loc[mut.obs['timepoint']=='52wk', :].sample(n=7230, replace=False).index.to_list()\n",
    "sample_mut = mut[(mut.obs.index.isin(mut26)|mut.obs.index.isin(mut52)), :]\n",
    "\n",
    "# save data objects\n",
    "sample_wt.write(os.path.join(dfp,'220224_WT_sampling.h5ad'))\n",
    "sample_mut.write(os.path.join(dfp,'220224_SBMA_sampling.h5ad'))\n",
    "print('saved @'+datetime.datetime.now().strftime('%y%m%d.%H:%M:%S'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WT 26wk\n",
      "OL     2924\n",
      "NU     2406\n",
      "AS2     877\n",
      "MG      293\n",
      "AS1     239\n",
      "OPC     220\n",
      "END     156\n",
      "EP       47\n",
      "MN       35\n",
      "CN       17\n",
      "PER      16\n",
      "Name: sub3, dtype: int64\n",
      " \n",
      "WT 52wk\n",
      "OL     2707\n",
      "NU     2486\n",
      "AS2    1054\n",
      "AS1     325\n",
      "OPC     251\n",
      "MG      232\n",
      "END     120\n",
      "EP       28\n",
      "CN       22\n",
      "MN        3\n",
      "PER       2\n",
      "Name: sub3, dtype: int64\n",
      " \n"
     ]
    }
   ],
   "source": [
    "for i in sample_wt.obs['timepoint'].unique():\n",
    "    wtt = sample_wt[sample_wt.obs['timepoint']==i,:]\n",
    "    print(\"WT \" + i)\n",
    "    print(wtt.obs['sub3'].value_counts())\n",
    "    print(\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBMA 26wk\n",
      "OL     2942\n",
      "NU     2391\n",
      "AS2     980\n",
      "MG      251\n",
      "OPC     224\n",
      "AS1     217\n",
      "END     118\n",
      "EP       42\n",
      "MN       24\n",
      "CN       24\n",
      "PER      17\n",
      "Name: sub3, dtype: int64\n",
      " \n",
      "SBMA 52wk\n",
      "OL     2767\n",
      "NU     2341\n",
      "AS2    1109\n",
      "AS1     351\n",
      "MG      267\n",
      "OPC     199\n",
      "END     110\n",
      "EP       53\n",
      "CN       18\n",
      "MN        9\n",
      "PER       6\n",
      "Name: sub3, dtype: int64\n",
      " \n"
     ]
    }
   ],
   "source": [
    "for i in sample_mut.obs['timepoint'].unique():\n",
    "    mutt = sample_mut[sample_mut.obs['timepoint']==i,:]\n",
    "    print(\"SBMA \" + i)\n",
    "    print(mutt.obs['sub3'].value_counts())\n",
    "    print(\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Evaluating 26wk\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "OL, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:2924\n",
      "    Ncells in Y:2942\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 28.83-s\n",
      "... 50% completed in 60.96-s\n",
      "... 75% completed in 97.55-s\n",
      "... mwu computed in 138.76-s\n",
      "\n",
      "... computed in 158.62-s\n",
      "\n",
      "AS2, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:877\n",
      "    Ncells in Y:980\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 20.25-s\n",
      "... 50% completed in 43.69-s\n",
      "... 75% completed in 71.53-s\n",
      "... mwu computed in 103.65-s\n",
      "\n",
      "... computed in 110.84-s\n",
      "\n",
      "NU, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:2406\n",
      "    Ncells in Y:2391\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 27.68-s\n",
      "... 50% completed in 58.52-s\n",
      "... 75% completed in 93.60-s\n",
      "... mwu computed in 132.81-s\n",
      "\n",
      "... computed in 146.39-s\n",
      "\n",
      "OPC, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:220\n",
      "    Ncells in Y:224\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.92-s\n",
      "... 50% completed in 37.03-s\n",
      "... 75% completed in 61.30-s\n",
      "... mwu computed in 89.49-s\n",
      "\n",
      "... computed in 93.58-s\n",
      "\n",
      "MG, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:293\n",
      "    Ncells in Y:251\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 17.16-s\n",
      "... 50% completed in 37.51-s\n",
      "... 75% completed in 62.19-s\n",
      "... mwu computed in 90.93-s\n",
      "\n",
      "... computed in 95.23-s\n",
      "\n",
      "MN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:35\n",
      "    Ncells in Y:24\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.90-s\n",
      "... 50% completed in 34.92-s\n",
      "... 75% completed in 58.17-s\n",
      "... mwu computed in 85.50-s\n",
      "\n",
      "... computed in 88.82-s\n",
      "\n",
      "AS1, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:239\n",
      "    Ncells in Y:217\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 17.13-s\n",
      "... 50% completed in 37.46-s\n",
      "... 75% completed in 62.23-s\n",
      "... mwu computed in 91.09-s\n",
      "\n",
      "... computed in 95.05-s\n",
      "\n",
      "END, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:156\n",
      "    Ncells in Y:118\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.36-s\n",
      "... 50% completed in 35.89-s\n",
      "... 75% completed in 59.61-s\n",
      "... mwu computed in 87.38-s\n",
      "\n",
      "... computed in 91.25-s\n",
      "\n",
      "PER, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:16\n",
      "    Ncells in Y:17\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.65-s\n",
      "... 50% completed in 34.50-s\n",
      "... 75% completed in 57.40-s\n",
      "... mwu computed in 84.37-s\n",
      "\n",
      "... computed in 87.60-s\n",
      "\n",
      "CN, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:17\n",
      "    Ncells in Y:24\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 16.24-s\n",
      "... 50% completed in 35.67-s\n",
      "... 75% completed in 59.19-s\n",
      "... mwu computed in 86.86-s\n",
      "\n",
      "... computed in 90.31-s\n",
      "\n",
      "EP, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:47\n",
      "    Ncells in Y:42\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 15.94-s\n",
      "... 50% completed in 35.13-s\n",
      "... 75% completed in 58.74-s\n",
      "... mwu computed in 86.47-s\n",
      "\n",
      "... computed in 89.84-s\n",
      "\n",
      "Finished timepoint 26wk in 19.13-min\n",
      "Evaluating 52wk\n",
      "\n",
      "--------\n",
      "...\n",
      "--------\n",
      "\n",
      "\n",
      "OL, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:2707\n",
      "    Ncells in Y:2767\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 28.43-s\n",
      "... 50% completed in 59.78-s\n",
      "... 75% completed in 95.30-s\n",
      "... mwu computed in 135.17-s\n",
      "\n",
      "... computed in 149.92-s\n",
      "\n",
      "AS2, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:1054\n",
      "    Ncells in Y:1109\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 20.76-s\n",
      "... 50% completed in 44.96-s\n",
      "... 75% completed in 73.14-s\n",
      "... mwu computed in 105.46-s\n",
      "\n",
      "... computed in 113.17-s\n",
      "\n",
      "NU, WT vs SBMA\n",
      "----\n",
      "    Ncells in X:2486\n",
      "    Ncells in Y:2341\n",
      "\n",
      "Mann-Whitney U w/Benjamini/Hochberg correction\n",
      "\n",
      "... 25% completed in 27.74-s\n",
      "... 50% completed in 58.75-s\n"
     ]
    }
   ],
   "source": [
    "## EMD; SUB3, IMP, Sample\n",
    "wt = sample_wt\n",
    "mut = sample_mut\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub3'\n",
    "    fname = 'imp_samp' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['26wk','52wk'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in wt.obs[group].unique() :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SBMA'.format(i))\n",
    "            print('----')\n",
    "            X = wt[((wt.obs[group]==i) & (wt.obs['timepoint']==t)), :].layers['imputed']\n",
    "            Y = mut[((mut.obs[group]==i) & (mut.obs['timepoint']==t)), :].layers['imputed']\n",
    "            \n",
    "\n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=adata.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,adata.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(X,Y,adata.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_down.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_up.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EMD; SUB3, No IMP,\n",
    "wt = sample_wt\n",
    "mut = sample_mut\n",
    "\n",
    "if True :\n",
    "    dge_grandtotal = time.time()\n",
    "    group='sub3'\n",
    "    fname = 'no_imp_samp' \n",
    "    dge = pd.DataFrame()\n",
    "    for t in ['26wk','52wk'] :\n",
    "        print('Evaluating {}'.format(t))\n",
    "        t_total = time.time()\n",
    "        dge_total = time.time()\n",
    "        start_t=time.time()\n",
    "        \n",
    "        # up down dichotomy\n",
    "        print('\\n--------')\n",
    "        print('...')\n",
    "        print('--------\\n')\n",
    "#        dge = pd.DataFrame()\n",
    "        for i in wt.obs[group].unique() :\n",
    "            start = time.time()\n",
    "            print('\\n{}, WT vs SBMA'.format(i))\n",
    "            print('----')\n",
    "            X = wt.X[((wt.obs[group]==i) & (wt.obs['timepoint']==t)), :]\n",
    "            Y = mut.X[((mut.obs[group]==i) & (mut.obs['timepoint']==t)), :]\n",
    "            \n",
    "            if sparse.issparse(X):\n",
    "                X = X.todense()\n",
    "            if sparse.issparse(Y):\n",
    "                Y = Y.todense()\n",
    "                \n",
    "            X = np.asarray(X)\n",
    "            Y = np.asarray(Y)\n",
    "        \n",
    "            print('    Ncells in X:{}'.format(X.shape[0]))\n",
    "            print('    Ncells in Y:{}\\n'.format(Y.shape[0]))            \n",
    "            \n",
    "            emd = scprep.stats.differential_expression(X,Y,\n",
    "                                                       measure = 'emd',\n",
    "                                                       direction='both',\n",
    "                                                       gene_names=adata.var_names,\n",
    "                                                       n_jobs=-1)\n",
    "            \n",
    "            # mann-whitney u, corrected p-values\n",
    "            p = mwu(X,Y,adata.var_names)\n",
    "            emd['Gene']=emd.index\n",
    "            emd=emd.drop(columns='rank')\n",
    "            fc = log2aveFC(X,Y,adata.var_names.to_list())\n",
    "            gene_mismatch = fc['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                fc = fc.loc[gene_mismatch,:]\n",
    "                warnings.warn('Warning: {} genes dropped due to p-val NA.'.format((gene_mismatch==False).sum()))\n",
    "            dt = pd.merge(p,fc,how='left',on=\"Gene\")\n",
    "            gene_mismatch = emd['Gene'].isin(p['Gene'])\n",
    "            if gene_mismatch.any():\n",
    "                emd = emd.loc[gene_mismatch,:]\n",
    "            dt = pd.merge(dt,emd,how='left',on='Gene')\n",
    "            dt['Cell type']=[i]*dt.shape[0]\n",
    "            dt['timepoint']=[str(t)]*dt.shape[0]\n",
    "            dt['nlog10pvalcorrected']=(-1)*np.log10(dt['pval_corrected'])\n",
    "            dge = dge.append(dt, ignore_index=True)\n",
    "            print('... computed in {:.2f}-s'.format(time.time()-start))\n",
    "        print('\\nFinished timepoint {} in {:.2f}-min'.format(t,(time.time()-start_t)/60))  \n",
    "    dgeup = dge.loc[dge['emd']>0,:] # take only 'up' (switch for down)\n",
    "    dgedown = dge.loc[dge['emd']<0,:] # take only 'down'\n",
    "    dgeup.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_down.csv'),index=False)\n",
    "    dgedown.to_csv(os.path.join(pfp,'dge_'+fname+'_SBMA_up.csv'),index=False)\n",
    "        \n",
    "\n",
    "    print('DGE finished in {:.2f}-min'.format((time.time()-dge_grandtotal)/60))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
